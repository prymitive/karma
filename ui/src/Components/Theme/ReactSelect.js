const ReactSelectColors = {
  Light: {
    color: "#fff",
    singleValueColor: "#000",
    backgroundColor: "#fff",
    borderColor: "#ced4da",
    focusedBoxShadow: "rgba(69, 90, 100, 0.25)",
    focusedBorderColor: "#819ba8",
    menuBackground: "#fff",
    optionHoverBackground: "#455a64",
    valueContainerBackground: "#fff",
    disabledValueContainerBackground: "#ecf0f1",
  },
  Dark: {
    color: "#fff",
    singleValueColor: "#fff",
    backgroundColor: "#fff",
    borderColor: "#444",
    focusedBoxShadow: "rgba(69, 90, 100, 0.25)",
    focusedBorderColor: "#819ba8",
    menuBackground: "#2b2b2b",
    optionHoverBackground: "#455a64",
    valueContainerBackground: "#444",
    disabledValueContainerBackground: "#303030",
  },
};

const ReactSelectStyles = (theme) => ({
  control: (base, state) =>
    state.isFocused
      ? {
          ...base,
          backgroundColor: theme.backgroundColor,
          outline: "0",
          outlineOffset: "-2px",
          boxShadow: `0 0 0 0.2rem ${theme.focusedBoxShadow}`,
          borderRadius: "0.25rem",
          borderColor: theme.focusedBorderColor,
          "&:hover": {
            borderColor: theme.focusedBorderColor,
          },
        }
      : {
          ...base,
          backgroundColor: "inherit",
          borderRadius: "0.25rem",
          borderColor: theme.borderColor,
          "&:hover": { borderColor: theme.borderColor },
        },
  valueContainer: (base, state) =>
    state.isMulti
      ? {
          ...base,
          borderTopLeftRadius: "0.25rem",
          borderBottomLeftRadius: "0.25rem",
          backgroundColor: state.isDisabled
            ? theme.disabledValueContainerBackground
            : theme.valueContainerBackground,
          paddingLeft: "4px",
          paddingRight: "4px",
          display: "flex",
          flexDirection: "row",
          flexWrap: "wrap",
          maxWidth: "100%",
          overflow: "hidden",
        }
      : {
          ...base,
          borderTopLeftRadius: "0.25rem",
          borderBottomLeftRadius: "0.25rem",
          backgroundColor: state.isDisabled
            ? theme.disabledValueContainerBackground
            : theme.valueContainerBackground,
        },
  singleValue: (base, state) => ({
    ...base,
    color: theme.singleValueColor,
  }),
  multiValue: (base, state) => ({
    ...base,
    borderRadius: "4px",
    backgroundColor: theme.optionHoverBackground,
    "&:hover": {
      backgroundColor: theme.optionHoverBackground,
    },
  }),
  multiValueLabel: (base, state) => ({
    ...base,
    color: theme.color,
    whiteSpace: "normal",
    wordWrap: "break-word",
    wordBreak: "break-word",
    "&:hover": {
      color: theme.color,
    },
  }),
  multiValueRemove: (base, state) => ({
    ...base,
    cursor: "pointer",
    color: theme.color,
    backgroundColor: "inherit",
    opacity: "0.4",
    borderRadius: "inherit",
    "&:hover": {
      color: theme.color,
      backgroundColor: "inherit",
      opacity: "0.75",
    },
  }),
  input: (base, state) => ({
    ...base,
    color: "inherit",
  }),
  indicatorsContainer: (base, state) => ({
    ...base,
    backgroundColor: state.isDisabled
      ? theme.disabledValueContainerBackground
      : theme.valueContainerBackground,
    borderTopRightRadius: "0.25rem",
    borderBottomRightRadius: "0.25rem",
  }),
  dropdownIndicator: (base, state) =>
    state.isFocused
      ? {
          ...base,
          "&:hover": { color: "inherit" },
        }
      : { ...base },
  menu: (base, state) => ({
    ...base,
    zIndex: 1500,
    backgroundColor: theme.menuBackground,
  }),
  option: (base, state) => ({
    ...base,
    color: "inherit",
    backgroundColor: "inherit",
    "&:hover": {
      color: theme.color,
      backgroundColor: theme.optionHoverBackground,
      cursor: "pointer",
    },
  }),
});

export { ReactSelectStyles, ReactSelectColors };
