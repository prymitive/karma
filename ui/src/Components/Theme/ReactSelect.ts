import { CSSObject } from "@emotion/serialize";
import {
  ControlProps,
  IndicatorContainerProps,
  OptionTypeBase,
  SingleValueProps,
  Styles,
  ValueContainerProps,
} from "react-select";

interface ReactSelectTheme {
  color: string;
  singleValueColor: string;
  backgroundColor: string;
  borderColor: string;
  focusedBoxShadow: string;
  focusedBorderColor: string;
  menuBackground: string;
  optionHoverBackground: string;
  valueContainerBackground: string;
  disabledValueContainerBackground: string;
}

interface ReactSelectThemes {
  Light: ReactSelectTheme;
  Dark: ReactSelectTheme;
}

const ReactSelectColors: ReactSelectThemes = {
  Light: {
    color: "#fff",
    singleValueColor: "#000",
    backgroundColor: "#fff",
    borderColor: "#ced4da",
    focusedBoxShadow: "rgba(69, 90, 100, 0.25)",
    focusedBorderColor: "#819ba8",
    menuBackground: "#fff",
    optionHoverBackground: "#455a64",
    valueContainerBackground: "#fff",
    disabledValueContainerBackground: "#ecf0f1",
  },
  Dark: {
    color: "#fff",
    singleValueColor: "#fff",
    backgroundColor: "#fff",
    borderColor: "#444",
    focusedBoxShadow: "rgba(69, 90, 100, 0.25)",
    focusedBorderColor: "#819ba8",
    menuBackground: "#2b2b2b",
    optionHoverBackground: "#455a64",
    valueContainerBackground: "#444",
    disabledValueContainerBackground: "#303030",
  },
};

interface StateFnT {
  isFocused: boolean;
  isMulti: boolean;
  isDisabled: boolean;
}

const ReactSelectStyles = <
  OptionType extends OptionTypeBase,
  IsMulti extends boolean
>(
  theme: ReactSelectTheme
): Styles<OptionType, IsMulti> => ({
  control: (base: CSSObject, props: ControlProps<OptionType, IsMulti>) =>
    props.isFocused
      ? {
          ...base,
          backgroundColor: props.isDisabled
            ? theme.disabledValueContainerBackground
            : theme.valueContainerBackground,
          outline: "0",
          outlineOffset: "-2px",
          boxShadow: `0 0 0 0.2rem ${theme.focusedBoxShadow}`,
          borderRadius: "0.25rem",
          borderColor: theme.focusedBorderColor,
          "&:hover": {
            borderColor: theme.focusedBorderColor,
          },
        }
      : {
          ...base,
          backgroundColor: props.isDisabled
            ? theme.disabledValueContainerBackground
            : theme.valueContainerBackground,
          borderRadius: "0.25rem",
          borderColor: theme.borderColor,
          "&:hover": { borderColor: theme.borderColor },
        },
  valueContainer: (
    base: CSSObject,
    props: ValueContainerProps<OptionType, IsMulti>
  ) =>
    props.isMulti
      ? {
          ...base,
          borderTopLeftRadius: "0.25rem",
          borderBottomLeftRadius: "0.25rem",
          paddingLeft: "4px",
          paddingRight: "4px",
          display: "flex",
          flexDirection: "row",
          flexWrap: "wrap",
          maxWidth: "100%",
          overflow: "hidden",
        }
      : {
          ...base,
          borderTopLeftRadius: "0.25rem",
          borderBottomLeftRadius: "0.25rem",
        },
  singleValue: (base: CSSObject, props: SingleValueProps<OptionType>) => ({
    ...base,
    color: theme.singleValueColor,
  }),
  multiValue: (base: CSSObject) => ({
    ...base,
    borderRadius: "4px",
    backgroundColor: theme.optionHoverBackground,
    "&:hover": {
      backgroundColor: theme.optionHoverBackground,
    },
  }),
  multiValueLabel: (base: CSSObject) => ({
    ...base,
    color: theme.color,
    whiteSpace: "normal",
    wordWrap: "break-word",
    wordBreak: "break-word",
    "&:hover": {
      color: theme.color,
    },
  }),
  multiValueRemove: (base: CSSObject) => ({
    ...base,
    cursor: "pointer",
    color: theme.color,
    backgroundColor: "inherit",
    opacity: "0.4",
    borderRadius: "inherit",
    "&:hover": {
      color: theme.color,
      backgroundColor: "inherit",
      opacity: "0.75",
    },
  }),
  input: (base: CSSObject) => ({
    ...base,
    color: "inherit",
  }),
  indicatorsContainer: (
    base: CSSObject,
    props: IndicatorContainerProps<OptionType, IsMulti>
  ) => ({
    ...base,
    backgroundColor: props.isDisabled
      ? theme.disabledValueContainerBackground
      : theme.valueContainerBackground,
    borderTopRightRadius: "0.25rem",
    borderBottomRightRadius: "0.25rem",
  }),
  dropdownIndicator: (base: CSSObject, state: StateFnT) =>
    state.isFocused
      ? {
          ...base,
          "&:hover": { color: "inherit" },
        }
      : { ...base },
  menu: (base: CSSObject) => ({
    ...base,
    zIndex: 1500,
    backgroundColor: theme.menuBackground,
  }),
  option: (base: CSSObject) => ({
    ...base,
    color: "inherit",
    backgroundColor: "inherit",
    "&:hover": {
      color: theme.color,
      backgroundColor: theme.optionHoverBackground,
      cursor: "pointer",
    },
  }),
});

export { ReactSelectStyles, ReactSelectColors };
