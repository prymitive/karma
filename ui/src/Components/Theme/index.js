import React, { useEffect } from "react";
import ReactDOM from "react-dom";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSun } from "@fortawesome/free-solid-svg-icons/faSun";

const DarkTheme = React.lazy(() => import("Styles/DarkTheme"));
const LightTheme = React.lazy(() => import("Styles/LightTheme"));

const Placeholder = () => {
  return ReactDOM.createPortal(
    <div
      style={{
        zIndex: 2000,
        backgroundColor: "#455a64",
        position: "fixed",
        top: 0,
        left: 0,
        width: "100vw",
        height: "100vh",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        flexDirection: "column",
      }}
    >
      <FontAwesomeIcon
        icon={faSun}
        size="lg"
        spin
        style={{
          color: "#5e7a88",
          fontSize: "14rem",
        }}
      />
    </div>,
    document.body
  );
};

const ThemeContext = React.createContext();

const BodyTheme = () => {
  const context = React.useContext(ThemeContext);

  useEffect(() => {
    document.body.classList.toggle("theme-dark", context.isDark);
    document.body.classList.toggle("theme-light", !context.isDark);
  }, [context.isDark]);

  return (
    <React.Suspense fallback={<Placeholder />}>
      {context.isDark ? <DarkTheme /> : <LightTheme />}
    </React.Suspense>
  );
};

export { BodyTheme, ThemeContext };
