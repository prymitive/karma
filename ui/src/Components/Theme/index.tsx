import React, { FC, useEffect } from "react";
import ReactDOM from "react-dom";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSun } from "@fortawesome/free-solid-svg-icons/faSun";

import { Styles } from "react-select";

const DarkTheme = React.lazy(() => import("Styles/DarkTheme"));
const LightTheme = React.lazy(() => import("Styles/LightTheme"));

const Placeholder = () => {
  return ReactDOM.createPortal(
    <div
      style={{
        zIndex: 2000,
        backgroundColor: "#455a64",
        position: "fixed",
        top: 0,
        left: 0,
        width: "100vw",
        height: "100vh",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        flexDirection: "column",
      }}
    >
      <FontAwesomeIcon
        icon={faSun}
        size="lg"
        spin
        style={{
          color: "#5e7a88",
          fontSize: "14rem",
        }}
      />
    </div>,
    document.body
  );
};

export interface ThemeCtx {
  isDark: boolean;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  reactSelectStyles: Styles<any, any>;
  animations: {
    duration: number;
  };
}

const ThemeContext = React.createContext({
  isDark: false,
  reactSelectStyles: {},
  animations: { duration: 1000 },
} as ThemeCtx);

const BodyTheme: FC = () => {
  const context = React.useContext(ThemeContext);

  useEffect(() => {
    document.body.classList.toggle("theme-dark", context.isDark);
    document.body.classList.toggle("theme-light", !context.isDark);
  }, [context.isDark]);

  return (
    <React.Suspense fallback={<Placeholder />}>
      {context.isDark ? <DarkTheme /> : <LightTheme />}
    </React.Suspense>
  );
};

export { BodyTheme, ThemeContext };
