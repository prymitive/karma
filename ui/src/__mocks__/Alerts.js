const MockAnnotation = (name, value, visible, isLink) => ({
  name: name,
  value: value,
  visible: visible,
  isLink: isLink,
});

const MockAlert = (annotations, labels, state) => ({
  id: Math.random().toString(36),
  annotations: annotations,
  labels: labels,
  startsAt: "2018-08-14T17:36:40.017867056Z",
  state: state,
  alertmanager: [
    {
      name: "default",
      cluster: "default",
      state: "active",
      startsAt: "2018-08-14T17:36:40.017867056Z",
      source: "localhost/prometheus",
      silencedBy: [],
      inhibitedBy: [],
    },
  ],
  receiver: "by-name",
});

const MockAlertGroup = (
  rootLabels,
  alerts,
  sharedAnnotations,
  sharedLabels,
  sharedSilences
) => ({
  receiver: "by-name",
  labels: rootLabels,
  alerts: alerts,
  id: "099c5ca6d1c92f615b13056b935d0c8dee70f18c",
  alertmanagerCount: {
    default: 1,
  },
  stateCount: {
    active: 1,
    suppressed: 0,
    unprocessed: 0,
  },
  shared: {
    annotations: sharedAnnotations,
    labels: sharedLabels,
    silences: sharedSilences,
  },
});

const MockSilence = () => ({
  comment: "Mocked Silence",
  createdAt: "0001-01-01T00:00:00Z",
  createdBy: "me@example.com",
  startsAt: "2000-01-01T00:00:00Z",
  endsAt: "2000-01-01T01:00:00Z",
  id: "04d37636-2350-4878-b382-e0b50353230f",
  ticketID: "",
  ticketURL: "",
  matchers: [
    { name: "foo", value: "bar", isRegex: false },
    { name: "baz", value: "regex", isRegex: true },
  ],
});

const MockAlertmanager = () => ({
  name: "default",
  cluster: "default",
  uri: "http://localhost",
  publicURI: "http://am.example.com",
  readonly: false,
  headers: {
    Authorization: "Basic foo bar",
  },
  corsCredentials: "include",
  error: "",
  version: "0.17.0",
  clusterMembers: ["default"],
});

export {
  MockAnnotation,
  MockAlert,
  MockAlertGroup,
  MockSilence,
  MockAlertmanager,
};
