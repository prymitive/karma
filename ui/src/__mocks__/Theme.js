import {
  ReactSelectStyles,
  ReactSelectColors,
} from "Components/Theme/ReactSelect";

const MockThemeContext = {
  animations: {
    in: undefined,
    duration: 500,
  },
  isDark: false,
  reactSelectStyles: ReactSelectStyles(ReactSelectColors.Light),
};

export { MockThemeContext };
