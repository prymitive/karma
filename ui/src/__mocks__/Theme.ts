import {
  ReactSelectStyles,
  ReactSelectColors,
} from "Components/Theme/ReactSelect";

const MockThemeContext = {
  animations: {
    duration: 500,
  },
  isDark: false,
  reactSelectStyles: ReactSelectStyles(ReactSelectColors.Light),
};

const MockThemeContextWithoutAnimations = {
  animations: {
    duration: 0,
  },
  isDark: false,
  reactSelectStyles: ReactSelectStyles(ReactSelectColors.Light),
};

export { MockThemeContext, MockThemeContextWithoutAnimations };
