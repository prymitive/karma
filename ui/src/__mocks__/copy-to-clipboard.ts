// mock copy-to-clipboard since it throws errors in tests
// and we don't really need to copy anything, only ensure we're calling it

const copy = jest.fn();

export default copy;
