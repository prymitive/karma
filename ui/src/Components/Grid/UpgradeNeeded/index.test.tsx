import React from "react";

import { mount, shallow } from "enzyme";

import toDiffableHtml from "diffable-html";

import { MockThemeContext } from "__mocks__/Theme";
import { UpgradeNeeded } from ".";

beforeEach(() => {
  jest.useFakeTimers();
  jest.clearAllTimers();
  jest.spyOn(React, "useContext").mockImplementation(() => MockThemeContext);
});

afterEach(() => {
  jest.clearAllTimers();
  jest.restoreAllMocks();
});

describe("<UpgradeNeeded />", () => {
  it("matches snapshot", () => {
    const tree = shallow(
      <UpgradeNeeded newVersion="1.2.3" reloadAfter={100000000} />
    );
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("calls window.location.reload after timer is done", () => {
    const reloadSpy = jest
      .spyOn(global.window.location, "reload")
      .mockImplementation(() => {});
    mount(<UpgradeNeeded newVersion="1.2.3" reloadAfter={100000000} />);
    jest.runOnlyPendingTimers();
    expect(reloadSpy).toBeCalled();
  });

  it("stops calling window.location.reload after unmount", () => {
    const reloadSpy = jest
      .spyOn(global.window.location, "reload")
      .mockImplementation(() => {});
    const tree = mount(
      <UpgradeNeeded newVersion="1.2.3" reloadAfter={100000000} />
    );
    tree.unmount();
    jest.runOnlyPendingTimers();
    expect(reloadSpy).not.toBeCalled();
  });
});
