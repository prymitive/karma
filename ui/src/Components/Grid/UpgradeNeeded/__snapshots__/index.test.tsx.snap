// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UpgradeNeeded /> matches snapshot 1`] = `
"
<h1 class=\\"display-1 text-placeholder screen-center\\">
  <div class=\\"container-fluid text-center\\">
    <div class=\\"shake-slow shake-constant mb-4\\">
      <svg
        aria-hidden=\\"true\\"
        focusable=\\"false\\"
        data-prefix=\\"fas\\"
        data-icon=\\"rocket\\"
        class=\\"svg-inline--fa fa-rocket screen-center-icon-big text-success\\"
        role=\\"img\\"
        xmlns=\\"http://www.w3.org/2000/svg\\"
        viewbox=\\"0 0 512 512\\"
      >
        <path
          fill=\\"currentColor\\"
          d=\\"M156.6 384.9L125.7 354c-8.5-8.5-11.5-20.8-7.7-32.2c3-8.9 7-20.5 11.8-33.8L24 288c-8.6 0-16.6-4.6-20.9-12.1s-4.2-16.7 .2-24.1l52.5-88.5c13-21.9 36.5-35.3 61.9-35.3l82.3 0c2.4-4 4.8-7.7 7.2-11.3C289.1-4.1 411.1-8.1 483.9 5.3c11.6 2.1 20.6 11.2 22.8 22.8c13.4 72.9 9.3 194.8-111.4 276.7c-3.5 2.4-7.3 4.8-11.3 7.2v82.3c0 25.4-13.4 49-35.3 61.9l-88.5 52.5c-7.4 4.4-16.6 4.5-24.1 .2s-12.1-12.2-12.1-20.9V380.8c-14.1 4.9-26.4 8.9-35.7 11.9c-11.2 3.6-23.4 .5-31.8-7.8zM384 168c22.1 0 40-17.9 40-40s-17.9-40-40-40s-40 17.9-40 40s17.9 40 40 40z\\"
        >
        </path>
      </svg>
    </div>
    <p class=\\"lead text-muted\\">
      <svg
        aria-hidden=\\"true\\"
        focusable=\\"false\\"
        data-prefix=\\"fas\\"
        data-icon=\\"spinner\\"
        class=\\"svg-inline--fa fa-spinner fa-spin me-2\\"
        role=\\"img\\"
        xmlns=\\"http://www.w3.org/2000/svg\\"
        viewbox=\\"0 0 512 512\\"
      >
        <path
          fill=\\"currentColor\\"
          d=\\"M304 48c0-26.5-21.5-48-48-48s-48 21.5-48 48s21.5 48 48 48s48-21.5 48-48zm0 416c0-26.5-21.5-48-48-48s-48 21.5-48 48s21.5 48 48 48s48-21.5 48-48zM48 304c26.5 0 48-21.5 48-48s-21.5-48-48-48s-48 21.5-48 48s21.5 48 48 48zm464-48c0-26.5-21.5-48-48-48s-48 21.5-48 48s21.5 48 48 48s48-21.5 48-48zM142.9 437c18.7-18.7 18.7-49.1 0-67.9s-49.1-18.7-67.9 0s-18.7 49.1 0 67.9s49.1 18.7 67.9 0zm0-294.2c18.7-18.7 18.7-49.1 0-67.9S93.7 56.2 75 75s-18.7 49.1 0 67.9s49.1 18.7 67.9 0zM369.1 437c18.7 18.7 49.1 18.7 67.9 0s18.7-49.1 0-67.9s-49.1-18.7-67.9 0s-18.7 49.1 0 67.9z\\"
        >
        </path>
      </svg>
      Upgrading to a new version: 1.2.3
    </p>
  </div>
</h1>
"
`;
